### ISSUE TITLE
Important Notes on 18.0.4snap4 SSL Compatibility With iOS/macOS devices

### ISSUE THREAD

**Issue by ctrezevant** [nextcloud/nextcloud-snap#1378]
Good afternoon, 

I'm writing to report that my iOS/macOS devices were initially unable to successfully negotiate an HTTPS connection with the latest version of the nextcloud snap. Based on my perusal of the repository's commit log, this was most likely introduced in e880e723ce8e25d5221b21ae27b51edf09962300. I have encountered this issue throughout macOS and iOS, with Safari and even the command line openssl-client being affected. 

This took me a bit of time to diagnose, so I'm sharing my solution here. Anyone facing this issue should take the following steps to resolve it:

First, If you're using ECC certs, make sure that they're either 256 or 384 bits. 521 bit keys based on curve secp521r1/P-521 are [widely deprecated](https://security.stackexchange.com/questions/100991/why-is-secp521r1-no-longer-supported-in-chrome-others) in various libraries and software packages.

You can get information about the size/curve of your public key using the following command:

```
openssl x509 -text -noout -in /path/to/your/certificate
``` 

If you see any of the following lines, you'll need to regenerate your certificate:

```
Public-Key: (521 bit)
ASN1 OID: secp521r1
NIST CURVE: P-521
```

Second, make sure you're including the [full certificate chain](https://whatsmychaincert.com/) when enabling HTTPS in the snap (via `nextcloud.enable-https`). In my particular case for example, I would include both the root and intermediary certificates of my internal SSL certificate authority.

If you have other questions or need additional clarification, feel free to comment on this issue.

**Comment by kyrofa** [comment]
When you say an issue was introduced by a given commit, you're typically saying that said commit introduced a regression. However, e880e723ce8e25d5221b21ae27b51edf09962300 is almost a year old (context for it is given [here](https://kyrofa.com/posts/cipher-update-for-the-nextcloud-snap)). It rather sounds like this issue is caused by generating your own cert and not configuring it properly, would you agree?

**Comment by ctrezevant** [comment]
It's partly both. While it was my mistake for generating a certificate using P-521 (which certain libraries decided to deprecate based on an [apparent misunderstanding of NSA Suite B](https://security.stackexchange.com/a/100995)), that same 521-bit certificate had been working properly in previous versions with protocols < TLSv1.2 enabled up to 18.04snap2. So it's not an issue with the snap itself (e880e723ce8e25d5221b21ae27b51edf09962300 is in line with current best practices anyways) as much as an edge case that can break people's SSL if their PKI is configured in a certain way. 

So yes, everything is fine with the snap itself and nothing needs to be fixed on your end. I only decided to open this issue because it's possible that another person might encounter the same problem, and wanted to provide a bit of useful context/documentation on how to resolve it.

Anyways, thanks again for doing such a great job with the snap- I've been using it for about a year now and it's always worked flawlessly. I really appreciate it!

**Comment by kyrofa** [comment]
Wait... the commit you referenced landed back in [15.0.5snap2](https://github.com/nextcloud/nextcloud-snap/blob/master/CHANGELOG#L105). The ciphers haven't changed since then. Nothing even closely related happened in 18.0.4snap3 or snap4 (check that CHANGELOG again). Are you sure this was working on 18.0.4snap2?

**Comment by ctrezevant** [comment]
That's odd- the container I have hosting the snap was created on October 18th of 2019, and I can see that the release for 15.0.5snap2 was [tagged](https://github.com/nextcloud/nextcloud-snap/releases/tag/15.0.5snap2) on Mar 24, 2019. 

What I do know is that I restored the container today from a snapshot I had taken on 5/27/20. I had changed the certificate to 2048 bit RSA late last week and experienced the SSL negotiation errors with that certificate (hence the restore), as well as with the old ECC certificate once the snapshot was restored and the snap auto-updated to 18.0.4snap4. The webpage (in Safari) and the iOS client only succeeded in reconnecting once I did a `snap revert` to 18.0.4snap2, but I have it permanently fixed in snap4 using the steps I outlined earlier. 

As far as how I set it up, I've always used the enable/disable-https commands and have never manually changed the apache configuration or enabled HSTS. That said, I never included a proper certificate chain in either case until now, so it's possible that was breaking the connections for my otherwise valid certs.
